Introduce EphemeralSource for symbols that should be simplified out

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113280
* #113279
* #118405
* __->__ #120948

Context: view fake-ification should handle closed-over state in ViewFuncs for use in view replay by:
* fake-ifying tensors
* symbolicizing SymInts

This avoids invalid specialization during view replay. However, the symbols / tensors created as intermediates in the view chain should not stick around or be guarded on. This PR introduces an `EphemeralSource` intended to be used as a source for this purpose. It has the following properties:
* Considered first to be simplified out in symbol simplification logic
* Errors if guarded on

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @aakhundov

Differential Revision: [D54561597](https://our.internmc.facebook.com/intern/diff/D54561597)

jbschlosser committed:
Introduce EphemeralSource for symbols that should be simplified out

[ghstack-poisoned]

ezyang commented on torch/_guards.py:
```diff
@@ -797,6 +801,9 @@ def is_dict_key(self):
         return self.base.is_dict_key()
 
+    def is_ephemeral(self):
+        return self.base.is_ephemeral()
```
is it even possible to chain an ephemeral source into a larger source 🤔

ezyang replied:
The answer is apparently yes, because ephemeral sources identify *tensors*, so you chain size/stride accessors off of them.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3661,8 +3661,20 @@ def _maybe_guard_eq(self, expr: Union["sympy.Eq", "sympy.Ne"], concrete_bool: bo
+        # Prefer to simplify out symbols with ephemeral sources.
+        def _smart_symbol_sort(x):
+            has_ephemeral_source = (
+                x in self.var_to_sources and any(s.is_ephemeral() for s in self.var_to_sources[x])
```
The any test here is kind of strange. If a symbol is associated with a non-ephemeral source, then all is fine, right? I've managed to unify the symbol with an actual input source and there is no reason to use it or not. So what you're worried about is the case when all of the sources are ephemeral (but in practice, you have exactly one source and it is ephemeral).

ezyang replied:
This also makes me wonder, if you end up with an EphemeralSource in `var_to_sources` with other stuff, you sure what to make sure you delete it later...

jbschlosser replied:
good points! changed the check to `all()` and added some logic to prefer non-ephemeral sources first in the `var_to_sources` list when possible, since the first ones are potentially used later on for guarding, etc.

ezyang commented on test/test_dynamic_shapes.py:
```diff
@@ -678,6 +681,51 @@ def test_statically_known_true(self):
+
+            def _get_ephemeral_source_symbols(t):
+                return [
+                    s.node._expr for s in (*t.shape, *t.stride(), t.storage_offset())
```
nit: itertools.chain avoids allocating the unpacked tuple and then iterating over it

ezyang commented on test/test_dynamic_shapes.py:
```diff
@@ -678,6 +681,51 @@ def test_statically_known_true(self):
+            def _get_ephemeral_source_symbols(t):
+                return [
+                    s.node._expr for s in (*t.shape, *t.stride(), t.storage_offset())
+                    if isinstance(s, torch.SymInt) and s.node._expr in shape_env.var_to_sources
```
are you sure about direct access to `_expr` here? I recently had to fix a bunch of code where they were accessing `_expr` and it was wrong, they should have accessed `expr`.

jbschlosser replied:
yeah I should definitely use `expr`. using that exhibits the behavior I originally expected where post-check none of the symints on x's metadata have ephemeral sources

ezyang commented on test/test_dynamic_shapes.py:
```diff
@@ -678,6 +681,51 @@ def test_statically_known_true(self):
+                    )
+                ]
+
+            # these checks should simplify out the ephemeral symbols, regardless of the
```
you should double check this test actually fails without your special sorter. I think it does, because y lexicographically sorts after x which means in reverse sort it would be preferentially picked first in all tie breaks... but if you had swapped the names then this test would not actually test what you want, as we would always consistently eliminate the ephemeral variable by accident (by virtue of its naming).

Actually, our heuristic where we try to substitute newer variables with older ones is pretty good, it probably is enough to eliminate ephemerals (except it doesn't actually work, because we string order rather than order on the allocation number LOL)

ezyang reviewed (approved):
okey dokey

ezyang commented:
To be honest, if I didn't have the overall context, it would be pretty hard to tell what was going on with this from just the comment on EphemeralSource

jbschlosser committed:
Update on "Introduce EphemeralSource for symbols that should be simplified out"


Context: view fake-ification should handle closed-over state in ViewFuncs for use in view replay by:
* fake-ifying tensors
* symbolicizing SymInts

This avoids invalid specialization during view replay. However, the symbols / tensors created as intermediates in the view chain should not stick around or be guarded on. This PR introduces an `EphemeralSource` intended to be used as a source for this purpose. It has the following properties:
* Considered first to be simplified out in symbol simplification logic
* Errors if guarded on

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce EphemeralSource for symbols that should be simplified out"


Context: view fake-ification should handle closed-over state in ViewFuncs for use in view replay by:
* fake-ifying tensors
* symbolicizing SymInts

This avoids invalid specialization during view replay. However, the symbols / tensors created as intermediates in the view chain should not stick around or be guarded on. This PR introduces an `EphemeralSource` intended to be used as a source for this purpose. It has the following properties:
* Considered first to be simplified out in symbol simplification logic
* Errors if guarded on

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54561597).
